dcm4che-imageio/src/main/java/org/dcm4che3/imageio/plugins/dcm/ParseEntireFileMetaData.java

```diff
@@ -0,0 +1,191 @@
+//                    All Rights Reserved                    //
+///////////////////////////////////////////////////////////////
+//                                                           //
+//       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF      //
+//        Agfa HealthCare N.V. and/or its affiliates.        //
+//      The copyright notice above does not evidence any     //
+//     actual or intended publication of such source code.   //
+//                                                           //
+///////////////////////////////////////////////////////////////
+//
+package org.dcm4che3.imageio.plugins.dcm;
+
+import org.dcm4che3.data.*;
+import org.dcm4che3.error.NoPixelDataException;
+import org.dcm4che3.image.PhotometricInterpretation;
+import org.dcm4che3.imageio.stream.BulkURIImageInputStream;
+import org.dcm4che3.imageio.stream.ByteArrayImageInputStream;
+import org.dcm4che3.io.stream.ImageInputStreamLoader;
+import org.dcm4che3.io.stream.ServiceImageInputStreamLoader;
+import org.dcm4che3.io.stream.DelegateImageInputStreamLoader;
+import org.dcm4che3.imageio.stream.ImageInputStreamAdapter;
+import org.dcm4che3.imageio.stream.SegmentedImageInputStream;
+import org.dcm4che3.io.BulkDataDescriptor;
+import org.dcm4che3.io.DicomInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.imageio.stream.ImageInputStream;
+import javax.imageio.stream.MemoryCacheImageInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.ByteOrder;
+import java.nio.file.Path;
+
+/**
+ * Load the entire file into memory as BulkData.  Pixels are streamed from the BulkData URI to clients.  This class
+ * sacrifices initial parse time for having a stateless class which is able to access the same file from different threads
+ * without synchronization.
+ *
+ */
+public class ParseEntireFileMetaData extends DicomMetaData {
+    private static final Logger LOG = LoggerFactory.getLogger(ParseEntireFileMetaData.class);
+
+    private static final ImageInputStreamLoader<Object> SERVICE_LOADER = new ServiceImageInputStreamLoader();
+
+    private final Attributes fmi;
+    private final Attributes attributes;
+    private final URI uri;
+    private final ImageInputStreamLoader<URI> uriLoader;
+
+    private final Object pixelData;
+    private final VR pixelDataVR;
+
+    // IIS that was used to load this file.
+
+
+    public ParseEntireFileMetaData(Object input) throws IOException {
+        this.uri = toURI(input);
+
+
+        String uriStr = this.uri.toString();
+        ImageInputStream iis = SERVICE_LOADER.openStream(input);
+        try (DicomInputStream dis = new DicomInputStream(new ImageInputStreamAdapter(iis))) {
+            dis.setIncludeBulkData(DicomInputStream.IncludeBulkData.URI);
+            dis.setBulkDataDescriptor(BulkDataDescriptor.PIXELDATA);
```

Comment:
It should be possible to provide the preferred bulk data descriptor, as PIXELDATA isn't really an optimal choice for large private tags data.

Reply:
The original implementation relied on people doing DicomMetaData subclasses to modify how the data was read.  I have addressed that in a number of ways i the latest build:  

- DicomMetaData is a concrete class again, and it defers the reading of attributes / values to DicomImageAccessor which is passed in the constructor.   This means that the Accessor can get data in any manner the application wants.
- If the user has the Attributes / FMI then they can be read in the way they want and passed through the constructor.
- The DefaultMetaDataFactory is a configurable factory which takes a URI and generates an appropriate DicomImageAccessor subclass.  One of the configurations on the factory is BulkDataDescriptor.